File: lingting-http/src/main/java/live/lingting/framework/http/HttpRequest.java
package live.lingting.framework.http;

import live.lingting.framework.http.body.BodySource;
import live.lingting.framework.http.body.MemoryBody;
import live.lingting.framework.http.header.HttpHeaders;
import live.lingting.framework.jackson.JacksonUtils;
import live.lingting.framework.util.StringUtils;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

import java.io.InputStream;
import java.net.URI;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Collection;
import java.util.Map;
import java.util.function.Consumer;

/**
 * @author lingting 2024-09-27 21:29
 */
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class HttpRequest {

	protected final HttpMethod method;

	protected final URI uri;

	protected final HttpHeaders headers;

	protected final Body body;

	public static Builder builder() {
		return new Builder();
	}

	public HttpMethod method() {
		return method;
	}

	public URI uri() {
		return uri;
	}

	public HttpHeaders headers() {
		return headers;
	}

	public Body body() {
		return body;
	}

	@Getter
	public static class Builder {

		private HttpMethod method = HttpMethod.GET;

		private HttpUrlBuilder urlBuilder;

		private final HttpHeaders headers = HttpHeaders.empty();

		private Body body;

		// region method
		public Builder method(HttpMethod method) {
			this.method = method;
			return this;
		}

		public Builder method(String method) {
			return method(HttpMethod.valueOf(method));
		}

		public Builder get() {
			return method(HttpMethod.GET);
		}

		public Builder put() {
			return method(HttpMethod.PUT);
		}

		public Builder post() {
			return method(HttpMethod.POST);
		}

		public Builder delete() {
			return method(HttpMethod.DELETE);
		}

		public Builder head() {
			return method(HttpMethod.HEAD);
		}

		public Builder options() {
			return method(HttpMethod.OPTIONS);
		}

		public Builder patch() {
			return method(HttpMethod.PATCH);
		}

		// endregion

		// region url
		public Builder url(String url) {
			return url(URI.create(url));
		}

		public Builder url(URI url) {
			this.urlBuilder = HttpUrlBuilder.from(url);
			return this;
		}

		public Builder url(Consumer<HttpUrlBuilder> consumer) {
			consumer.accept(urlBuilder);
			return this;
		}

		public Builder url(HttpUrlBuilder urlBuilder) {
			this.urlBuilder = urlBuilder;
			return this;
		}

		// endregion

		// region headers

		public Builder header(String name, String value) {
			this.headers.add(name, value);
			return this;
		}

		public Builder headers(String name, Collection<String> values) {
			this.headers.addAll(name, values);
			return this;
		}

		public Builder headers(Map<String, Collection<String>> map) {
			this.headers.addAll(map);
			return this;
		}

		public Builder headers(HttpHeaders headers) {
			this.headers.addAll(headers);
			return this;
		}

		public Builder headers(Consumer<HttpHeaders> consumer) {
			consumer.accept(headers);
			return this;
		}

		// endregion

		// region body

		public Builder body(Body body) {
			this.body = body;
			String contentType = body.contentType();
			headers.contentType(contentType);
			return this;
		}

		public Builder body(BodySource body) {
			return body(Body.of(body, headers.contentType()));
		}

		public Builder body(String body) {
			return body(new MemoryBody(body));
		}

		public Builder jsonBody(Object obj) {
			return body(JacksonUtils.toJson(obj));
		}

		// endregion

		public HttpRequest build() {
			Body requestBody = body == null ? Body.empty() : body;
			String contentType = headers.contentType();
			if (!StringUtils.hasText(contentType)) {
				String type = requestBody.contentType();
				if (StringUtils.hasText(type)) {
					headers.contentType(type);
				}
			}
			URI uri = urlBuilder.buildUri();
			headers.host(uri.getHost());
			return new HttpRequest(method, uri, headers.unmodifiable(), requestBody);
		}

	}

	/**
	 * @author lingting 2024-09-28 11:54
	 */
	@RequiredArgsConstructor
	public static class Body {

		private final BodySource source;

		private final String contentType;

		public static Body empty() {
			return new Body(BodySource.empty(), null);
		}

		public static Body of(BodySource body) {
			return new Body(body, null);
		}

		public static Body of(BodySource body, String contentType) {
			return new Body(body, contentType);
		}

		public String contentType() {
			return contentType;
		}

		public long contentLength() {
			return source.length();
		}

		public BodySource source() {
			return source;
		}

		public byte[] bytes() {
			return source.bytes();
		}

		public InputStream input() {
			return source.openInput();
		}

		public String string() {
			return string(StandardCharsets.UTF_8);
		}

		@SneakyThrows
		public String string(Charset charset) {
			return source.string(charset);
		}

	}

}

File: lingting-http/src/main/kotlin/live/lingting/framework/http/HttpRequest.kt
package live.lingting.framework.http

import live.lingting.framework.http.body.Body
import live.lingting.framework.http.body.MemoryBody
import live.lingting.framework.http.body.RequestBody
import live.lingting.framework.http.header.HttpHeaders
import live.lingting.framework.jackson.JacksonUtils
import java.net.URI
import java.util.function.Consumer

/**
 * @author lingting 2024-09-27 21:29
 */
open class HttpRequest private constructor(
    protected val method: HttpMethod,
    protected val uri: URI,
    protected val headers: HttpHeaders,
    protected val body: RequestBody
) {

    companion object {
        @JvmStatic
        fun builder(): Builder {
            return Builder()
        }
    }

    fun method(): HttpMethod {
        return method
    }

    fun uri(): URI {
        return uri
    }

    fun headers(): HttpHeaders {
        return headers
    }

    fun body(): RequestBody {
        return body
    }

    override fun toString(): String {
        return "[$method] $uri"
    }

    class Builder {
        var method: HttpMethod = HttpMethod.GET
            private set

        var urlBuilder: HttpUrlBuilder = HttpUrlBuilder()
            private set

        val headers: HttpHeaders = HttpHeaders.empty()

        var body: RequestBody = RequestBody.empty()
            private set

        // region method
        fun method(method: HttpMethod): Builder {
            this.method = method
            return this
        }

        fun method(method: String): Builder {
            return method(HttpMethod.valueOf(method))
        }

        fun get(): Builder {
            return method(HttpMethod.GET)
        }

        fun put(): Builder {
            return method(HttpMethod.PUT)
        }

        fun post(): Builder {
            return method(HttpMethod.POST)
        }

        fun delete(): Builder {
            return method(HttpMethod.DELETE)
        }

        fun head(): Builder {
            return method(HttpMethod.HEAD)
        }

        fun options(): Builder {
            return method(HttpMethod.OPTIONS)
        }

        fun patch(): Builder {
            return method(HttpMethod.PATCH)
        }

        // endregion

        // region url
        fun url(url: String): Builder {
            return url(HttpUrlBuilder.from(url))
        }

        fun url(url: URI): Builder {
            this.urlBuilder = HttpUrlBuilder.from(url)
            return this
        }

        fun url(consumer: Consumer<HttpUrlBuilder>): Builder {
            consumer.accept(urlBuilder)
            return this
        }

        fun url(urlBuilder: HttpUrlBuilder): Builder {
            this.urlBuilder = urlBuilder
            return this
        }

        // endregion

        // region headers
        fun header(name: String, value: String): Builder {
            headers.add(name, value)
            return this
        }

        fun headers(name: String, values: Collection<String>): Builder {
            headers.addAll(name, values)
            return this
        }

        fun headers(map: Map<String, Collection<String>>): Builder {
            headers.addAll(map)
            return this
        }

        fun headers(headers: HttpHeaders): Builder {
            this.headers.addAll(headers)
            return this
        }

        fun headers(consumer: Consumer<HttpHeaders>): Builder {
            consumer.accept(headers)
            return this
        }

        // endregion

        // region body
        fun body(body: RequestBody): Builder {
            this.body = body
            val contentType = body.contentType()
            headers.contentType(contentType)
            return this
        }

        fun body(body: Body): Builder {
            return body(RequestBody.of(body, headers.contentType()))
        }

        fun body(body: String): Builder {
            return body(MemoryBody(body))
        }

        fun jsonBody(obj: Any): Builder {
            return body(JacksonUtils.toJson(obj))
        }

        // endregion

        fun build(): HttpRequest {
            val requestBody = body
            // 不覆盖已有host
            val host = headers.host()
            if (host.isNullOrBlank()) {
                headers.host(urlBuilder.headerHost())
            }
            // 不覆盖已有contentType
            val contentType = headers.contentType()
            if (contentType.isNullOrBlank()) {
                val type = requestBody.contentType()
                if (!type.isNullOrBlank()) {
                    headers.contentType(type)
                }
            }
            val uri = urlBuilder.buildUri()
            return HttpRequest(method, uri, headers.unmodifiable(), requestBody)
        }
    }

}
